### PR TITLE
Implements Offline APO

### PR DESCRIPTION
**Pull request by jdchang1**
Main Changes:
- Added non-preference based offline RL track
- Offline Dataloader
- Offline Model (HF/MPT)
- Offline forward / loss

Backwards compatibility breaking change
`offline_rl` callback now refers to the single prompt/response case
`pairwise_offline_rl` callback is what was previously `offline_rl`

Working runs are:
`apo-single-stream-openr1-gBSwx1`

### CODE REVIEW

File: compose_rl/algorithms/offline/model_methods.py [hunk 1]
```diff
@@ -35,6 +39,114 @@ class PairwiseOfflineEnum(Enum):
...
+        policy_model_config: Policy model config.
+    """
+    if policy_model_config is not None and hasattr(model, 'transformer'):
+        clear_mb_load_balancing_loss(
+            policy_model_config,
+            model.transformer,  # type: ignore
+        )
+
+    # If we can't use attn_seq_id then we need to unpack each batch and
+    # Pack along the batch dimension instead.
```
Comment (bcui-db) [inline compose_rl/algorithms/offline/model_methods.py:65]:
```suggestion
```

File: compose_rl/algorithms/offline/model_methods.py [hunk 2]
```diff
@@ -35,6 +39,114 @@ class PairwiseOfflineEnum(Enum):
...
+                torch.clip(policy_logp - ref_logp, max=5.),
+            ) * batch['reward']
+            estimated_reward = torch.mean(estimated_rewards)
+
+    losses = losses.mean()
+
+    implicit_rewards = beta * (policy_logp - ref_logp).detach()
+
+    # Logging KL margins for comparing different methods
+    reverse_kl = (policy_logp - ref_logp).detach()
```
Comment (bcui-db) [inline compose_rl/algorithms/offline/model_methods.py:131]: Would you want to use the `approx_kl` util here:

https://github.com/databricks/compose-rl/blob/1f6122b0dbb6432f55df6f03dd1cf8b3e9d699e4/compose_rl/utils/utils.py#L73C5-L73C14

File: compose_rl/data/offline_data.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    rewards = []
+    vstars = []
+
+    for sample in data:
+        input_ids = sample['input_ids']
+        prompt_len = sample['prompt_len']
+        sequence_len = sample['sequence_len']
+
+        # Note: if we do any truncation, we force the last token to be EOS
+        # https://github.com/mosaicml/RLHF/issues/101
```
Comment (bcui-db) [inline compose_rl/data/offline_data.py:53]: Probably modify the link to this issue?

File: compose_rl/data/offline_data.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+        # Add the eos token if it's not in the chosen sample
+        if input_ids[-1] != tokenizer.eos_token_id:
+            input_ids[-1] = tokenizer.eos_token_id  # type: ignore
+
+        pad_len = max_seq_len - sequence_len
+
+        if pad_len < 0:
+            # We should truncate with an additional token left for eos
+            truncate_len = pad_len + 1
```
Comment (bcui-db) [inline compose_rl/data/offline_data.py:63]:
```suggestion
            truncate_len = abs(pad_len) + 1
```

File: compose_rl/data/preference_data.py [hunk 5]
```diff
@@ -54,6 +54,7 @@ def pairwise_preference_dataset_collate_fn(
     prompt_lens = []
     sequence_id = []
     chosen_rewards = []
+    vstars = []
```
Comment (bcui-db) [inline compose_rl/data/preference_data.py:57]: Did you mean to modify this file?

**Review by bcui-db — COMMENTED** [review]
Mostly seems good. Left some comments around some nits, seems like extra code might've accidentally added. 

Also can we add some unit tests just for E2E testing of the new algorithm/model class?
